nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/iot/AbstractAWSIoTProcessor.java

```diff
@@ -0,0 +1,230 @@
+import org.apache.commons.lang3.RandomStringUtils;
+import org.apache.nifi.annotation.lifecycle.OnStopped;
+import org.apache.nifi.components.PropertyDescriptor;
+import org.apache.nifi.processor.ProcessContext;
+import org.apache.nifi.processor.util.StandardValidators;
+import org.apache.nifi.processors.aws.AbstractAWSCredentialsProviderProcessor;
+import org.apache.nifi.processors.aws.iot.util.AWS4Signer;
+import org.apache.nifi.processors.aws.iot.util.MqttWebSocketAsyncClient;
+import org.eclipse.paho.client.mqttv3.MqttConnectOptions;
+import org.eclipse.paho.client.mqttv3.MqttException;
+
+import java.util.Date;
+import java.util.concurrent.TimeUnit;
+
+public abstract class AbstractAWSIoTProcessor extends AbstractAWSCredentialsProviderProcessor<AWSIotClient> {
+    static final String PROP_NAME_ENDPOINT = "aws.iot.endpoint";
+    static final String PROP_NAME_CLIENT = "aws.iot.mqtt.client";
+    static final String PROP_NAME_KEEPALIVE = "aws.iot.mqtt.keepalive";
+    static final String PROP_NAME_TOPIC = "aws.iot.mqtt.topic";
+    static final String PROP_NAME_QOS = "aws.iot.mqtt.qos";
+    /**
+     * Amazon's current service limit on websocket connection duration
+     */
+    static final Integer PROP_DEFAULT_KEEPALIVE = 60 * 60 * 24;
+    /**
+     * When to start indicating the need for connection renewal (in seconds before actual termination)
+     */
+    static final Integer DEFAULT_CONNECTION_RENEWAL_BEFORE_KEEP_ALIVE_EXPIRATION = 20;
+    static final String PROP_DEFAULT_CLIENT = AbstractAWSIoTProcessor.class.getSimpleName();
+    /**
+     * Default QoS level for message delivery
+     */
+    static final Integer DEFAULT_QOS = 0;
+    String awsTopic;
+    int awsQos;
+    MqttWebSocketAsyncClient mqttClient;
+    String awsEndpoint;
+    String awsClientId;
+
+    private String awsRegion;
+    private Integer awsKeepAliveSeconds;
+    private Date dtLastConnect;
+
+    public static final PropertyDescriptor PROP_ENDPOINT = new PropertyDescriptor
+            .Builder().name(PROP_NAME_ENDPOINT)
+            .description("Your endpoint identifier in AWS IoT (e.g. A1B71MLXKNCXXX)")
+            .required(true)
+            .addValidator(StandardValidators.NON_EMPTY_VALIDATOR)
+            .build();
+
+    public static final PropertyDescriptor PROP_CLIENT = new PropertyDescriptor
+            .Builder().name(PROP_NAME_CLIENT)
+            .description("MQTT client ID to use. Under the cover your input will be extended by a random " +
+                    "string to ensure a unique id among all conntected clients.")
+            .required(false)
+            .defaultValue(PROP_DEFAULT_CLIENT)
+            .addValidator(StandardValidators.NON_EMPTY_VALIDATOR)
+            .build();
+
+    public static final PropertyDescriptor PROP_KEEPALIVE = new PropertyDescriptor
+            .Builder().name(PROP_NAME_KEEPALIVE)
+            .description("Seconds a WebSocket-connection remains open after automatically renewing it. " +
```

Comment:
Little confusing, maybe should read "The number of seconds a WebSocket-connection..."